pycbc/psd/__init__.py

```diff
@@ -72,25 +72,40 @@ def from_cli(opt, length, delta_f, low_frequency_cutoff,
     except AttributeError:
         psd_estimation = False
 
-    if (opt.psd_model or opt.psd_file or opt.asd_file)\
-                      and not psd_estimation:
+    exclusive_opts = [opt.psd_model, opt.psd_file, opt.asd_file,
+                      psd_estimation]
+    if sum(map(bool, exclusive_opts)) != 1:
+        err_msg = "You must specify exactly one of '--psd-file', "
+        err_msg += "'--psd-model', '--asd-file', '--psd-estimation'"
+        raise ValueError(err_msg)
+
+    if (opt.psd_model or opt.psd_file or opt.asd_file):
         # PSD from lalsimulation or file
         if opt.psd_model:
             psd = from_string(opt.psd_model, length, delta_f, f_low)
-        elif opt.psd_file:
-            psd = from_txt(opt.psd_file, length,
-                           delta_f, f_low, is_asd_file=False)
-        elif opt.asd_file:
-            psd = from_txt(opt.asd_file, length,
-                           delta_f, f_low, is_asd_file=True)
+        elif opt.psd_file or opt.asd_file:
+            if opt.asd_file:
+                psd_file_name = opt.asd_file
+            else:
+                psd_file_name = opt.psd_file
+            if psd_file_name.endswith(('.dat', '.txt')):
+                is_asd_file = bool(opt.asd_file)
+                psd = from_txt(psd_file_name, length,
+                               delta_f, f_low, is_asd_file=is_asd_file)
+            elif opt.asd_file:
+                err_msg = "ASD files are only valid as ASCII files (.dat or "
+                err_msg += ".txt). Supplied {}.".format(psd_file_name)
+            elif psd_file_name.endswith(('.xml', '.xml.gz')):
+                psd = from_xml(psd_file_name, length, delta_f, f_low,
+                               ifo_string=opt.psd_file_xml_ifo_string,
+                               root_name=opt.psd_file_xml_root_name)
```

Comment:
Do the defaults make consistent sense here? I.E. if I don't give the ifo string or root name options, these options will both get "None". The default for root_name in particular seems to be 'psd' in your function. Does it make sense to make the defaults consistent?

Reply:
How do you mean? The default value for root_name is "psd" from the option parser? .... Or am I missing something?

Reply:
Argh, you are right! Sorry, I somehow didn't see the default in the argument parser. Ok, I think this is all fine then, except for the issue raised by coday.